python/taichi/lang/matrix.py

```diff
@@ -138,52 +138,39 @@ def is_global(self):
                 "Matrices with mixed global/local entries are not allowed"
         return results[0]
 
-    def is_pyconstant(self):
-        return all([not isinstance(e, expr.Expr) for e in self.entries])
-
-    @staticmethod
-    def make_from_numpy(nparray):
-        return Matrix(nparray)
-
-    @taichi_scope
     def element_wise_binary(self, foo, other):
         ret = self.empty_copy()
         if isinstance(other, (list, tuple)):
             other = Matrix(other)
         if foo.__name__ == 'assign' and not isinstance(other, Matrix):
             raise SyntaxError(
                 'cannot assign scalar expr to '
-                f'taichi class {type(a)}, maybe you want to use `a.fill(b)` instead?'
+                f'taichi class {type(self)}, maybe you want to use `a.fill(b)` instead?'
             )
         if isinstance(other, Matrix):
             assert self.m == other.m and self.n == other.n, f"Dimension mismatch between shapes ({self.n}, {self.m}), ({other.n}, {other.m})"
             for i in range(self.n * self.m):
                 ret.entries[i] = foo(self.entries[i], other.entries[i])
         else:  # assumed to be scalar
-            other = expr.Expr(other)
             for i in range(self.n * self.m):
                 ret.entries[i] = foo(self.entries[i], other)
         return ret
 
-    @taichi_scope
     def element_wise_unary(self, foo):
         ret = self.empty_copy()
         for i in range(self.n * self.m):
             ret.entries[i] = foo(self.entries[i])
         return ret
 
     def __matmul__(self, other):
-        # TODO: move to common_ops.py, redirect to `ti.matmul` too?
-        if self.is_pyconstant() and in_python_scope():
-            return self.make_from_numpy(self.to_numpy() @ other.to_numpy())
-
         assert self.m == other.n, f"Dimension mismatch between shapes ({self.n}, {self.m}), ({other.n}, {other.m})"
-        ret = Matrix(self.n, other.m)
+        ret = Matrix.empty(self.n, other.m)
         for i in range(self.n):
             for j in range(other.m):
-                ret(i, j).assign(self(i, 0) * other(0, j))
+                acc = self(i, 0) * other(0, j)
                 for k in range(1, other.n):
-                    ret(i, j).assign(ret(i, j) + self(i, k) * other(k, j))
+                    acc = acc + self(i, k) * other(k, j)
+                ret.set_entry(i, j, acc)
```

Comment:
Is `set_entry` same as `assign` in Taichi-scope?

Reply:
No, thank for pointing out, should we make `set_entry` call `assign` in Taichi-scope and `=` in Python-scope? Not sure this is a good design, but it's the most strightforward makes matmul work in Python-scope, which is the PR aim.

Reply:
Yeah I think that's a good idea for now. At least it makes me feel safer since we are not changing Taichi-space behavior.

Reply:
Thank for approving my idea. BTW I just come up with a more crazy idea: all exprs are not-assignable, just like LLVM IR does, which permanantly prevent such issues when doing `a = a.transpose()` or more complicated assignment dependency issues (iapr of course).

Reply:
We still need to have mutable local variables, so I'm afraid that won't work very well.